readme.md

```diff
@@ -1 +1,73 @@
-http://todomvc.com but for back end.
+# Goal
+
+* Create various implementations of a minimally simple addressbook service
+
+# Why?
+
+* To ease comparison between different languages/frameworks/approaches
+* To demonstrate how to use the various languages/frameworks/approaches
+
+## The Original Idea
+
+http://todomvc.com but using server-side technologies.
+
+* have a database with a table with names and phone numbers (and probably a surrogate key)
+* the application should list all the entries (optionally paged)
+* you should be able to add, edit and delete an entry
+* each entry has 2 fields: name and phone number
+* *any* other features are optional
+
+
+# Implementation Guidelines
+
+## General
+
+* Go for clarity and simplicity which should ease comparison and improve the pedagogical win.
+* Err on the side of excessive commenting.
+* Implementations MUST include persistence.
+* Implementations SHOULD behave as described in the API doc below, *whether it provides an actual API service or merely outputs HTML directly*.
+
+## Contacts
+
+A contact is completely described as:
+
+```json
+{
+  "id":   <integer>,
+  "name": "<string>",
+  "phone": "<string>"
+}
+```
+
+### SQL
+
+If you choose to use an SQL-based persistence, here is some example [DDL]:
+
+```sql
+CREATE TABLE contacts (
+id INTEGER NOT NULL PRIMARY UNIQUE AUTOINCREMENT,
+name TEXT NOT NULL,
+phone TEXT NOT NULL
+);
+```
+
+The above is SQLite compatible syntax.
+
+### File
+
+If you choose to use file-based persistence, please use OS-independent path handling/access where possible.
+
+## API
+
+|        URI        | Responses[1]  |   HTTP Verb/Parameters                        |        Description                  |
+|:------------------|---------------|-----------------------------------------------|------------------------------------:|
+| /contacts         | 200, 204      | none                                          | list all contacts                   |
+| /contact          | 201, 400      | PUT with JSON representation of new contact   | create a new contact                |
+| /contact/\<id>    | 200, 404      | GET                                           | display a specific contact          |
+| /contact/\<id>    | 200, 400, 404 | POST with JSON representation of *new* data   | update a specific contact           |
+| /contact/\<id>    | 204, 404      | DELETE                                        | delete a specific contact           |
+
```

Comment:
These endpoints seem pretty good but my implementation is non-compliant as of this writing. I have stuff like:

```
- GET people/list
- POST people
- DELETE people/<id>
```

https://github.com/pdurbin/addressbookmvc/blob/890dbdb6a96a10430350dcf533a605a4ad317bf9/examples/javaee7/src/test/java/com/greptilian/addressbookmvc/javaee7/api/PeopleIT.java

@prologic seems to be using "/addresses" as his endpoint: https://github.com/pdurbin/addressbookmvc/blob/288f97f27893aaf7542ef43d5de2e7f775312069/examples/circuits/server.py . I don't see a way to edit via PUT, by the way.

(@dhutty shouldn't POST be add and PUT be edit?)

Even though the API is completely optional, I feel a little stronger about it being the same than I do about all the database table and field names and types being identical.